File: src/game/transport/pathfinding.rs
use bevy::prelude::*;
use std::cmp::Ordering;
use std::collections::{BinaryHeap, HashMap, VecDeque};

use crate::game::intersections::IntersectionIndex;
use crate::game::map::{MapGrid, TilePos};
use crate::game::roads::RoadDir;
use crate::game::traffic::TrafficOccupancy;

use super::{RegionGraph, RoadGraph};

mod cost;

#[derive(Resource, serde::Serialize, serde::Deserialize, Debug, Clone)]
pub struct PathfindingConfig {
    pub cache_capacity: usize,
    pub cache_ttl_secs: f64,
    /// Congestion weight factor `k` in: `w = base_cost * (1 + k * congestion)`.
    pub congestion_k: f32,
    /// Clamp for congestion ratio (occupancy / capacity).
    pub congestion_max: f32,
    /// Cost penalty for lane changes (perpendicular move, same travel dir).
    pub lane_change_penalty: f32,
    /// Cost penalty for turns (perpendicular move into new dir lane).
    pub turn_penalty: f32,
    /// Scale factor converting float weights into integer A* costs.
    pub cost_scale: f32,

    /// Enable a simple hierarchical pre-pass (region graph) to prune A* search space.
    pub enable_hierarchical: bool,
    /// Region size in tiles for the hierarchical pre-pass.
    pub region_size: usize,
    /// How many neighbor regions around the high-level path to include (safety margin).
    pub region_pad: i32,
}

impl Default for PathfindingConfig {
    fn default() -> Self {
        Self {
            cache_capacity: 4096,
            cache_ttl_secs: 10.0,
            congestion_k: 2.0,
            congestion_max: 2.0,
            lane_change_penalty: 40.0,
            turn_penalty: 80.0,
            cost_scale: 1000.0,

            enable_hierarchical: true,
            region_size: 16,
            region_pad: 1,
        }
    }
}

#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
struct PathKey {
    start: TilePos,
    goal: TilePos,
    version: u64,
}

#[derive(Clone, Debug)]
struct CacheEntry {
    path: Vec<TilePos>,
    last_used_sec: f64,
}

/// A very small, dependency-free cache (TTL + approximate LRU).
#[derive(Resource, Debug, Default)]
pub struct PathCache {
    map: HashMap<PathKey, CacheEntry>,
    // (key, last_used_sec_at_push) allows skipping stale deque entries
    lru: VecDeque<(PathKey, f64)>,
}

impl PathCache {
    pub fn clear(&mut self) {
        self.map.clear();
        self.lru.clear();
    }
}

/// Find a road path using the current road graph and cache.
///
/// Returns empty vec if:
/// - graph not built for version
/// - start/goal not road nodes
/// - no path found
pub struct PathfindingCtx<'a> {
    pub time_now_sec: f64,
    pub cfg: &'a PathfindingConfig,
    pub cache: &'a mut PathCache,
    pub graph: &'a RoadGraph,
    pub regions: Option<&'a RegionGraph>,
    pub traffic: &'a TrafficOccupancy,
    pub grid: &'a MapGrid,
    pub intersections: &'a IntersectionIndex,
}

#[derive(Copy, Clone, Eq, PartialEq)]
struct HeapState {
    f: u32,
    g: u32,
    idx: usize,
}

impl Ord for HeapState {
    fn cmp(&self, other: &Self) -> Ordering {
        // reverse for min-heap behavior
        other
            .f
            .cmp(&self.f)
            .then_with(|| other.g.cmp(&self.g))
            .then_with(|| other.idx.cmp(&self.idx))
    }
}
impl PartialOrd for HeapState {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

fn manhattan_idx(a: usize, b: usize, w: usize) -> u32 {
    let ax = (a % w) as i32;
    let ay = (a / w) as i32;
    let bx = (b % w) as i32;
    let by = (b / w) as i32;
    ax.abs_diff(bx) + ay.abs_diff(by)
}

fn idx_to_pos(idx: usize, w: usize) -> TilePos {
    TilePos {
        x: (idx % w) as i32,
        y: (idx / w) as i32,
    }
}

pub fn find_road_path_cached(
    ctx: &mut PathfindingCtx<'_>,
    start: TilePos,
    goal: TilePos,
) -> Vec<TilePos> {
    if start == goal {
        return vec![start];
    }
    if ctx.graph.edges.is_empty() || ctx.graph.width == 0 {
        return Vec::new();
    }

    let w = ctx.graph.width;
    let len = ctx.graph.edges.len();

    let start_idx = (start.y as isize) * (w as isize) + (start.x as isize);
    let goal_idx = (goal.y as isize) * (w as isize) + (goal.x as isize);
    if start_idx < 0 || goal_idx < 0 {
        return Vec::new();
    }
    let start_idx = start_idx as usize;
    let goal_idx = goal_idx as usize;
    if start_idx >= len || goal_idx >= len {
        return Vec::new();
    }
    if ctx.graph.edges[start_idx] == 0 {
        // Must be on road graph (note: dead-end road would still have mask != 0 if connected).
        return Vec::new();
    }

    let key = PathKey {
        start,
        goal,
        version: ctx.graph.version,
    };

    // TTL eviction on access
    if let Some(entry) = ctx.cache.map.get(&key)
        && ctx.time_now_sec - entry.last_used_sec <= ctx.cfg.cache_ttl_secs
    {
        // refresh LRU
        ctx.cache.lru.push_back((key, ctx.time_now_sec));
        // clone path out (small vectors in MVP; ok)
        return entry.path.clone();
    }

    let allowed_regions = compute_allowed_regions(ctx, start, goal);

    // Try pruned search first, then fall back to full A* if needed.
    for attempt in 0..2 {
        let allowed = if attempt == 0 {
            allowed_regions.as_deref()
        } else {
            None
        };

        if let Some(out) = astar_road_graph(ctx, start_idx, goal_idx, allowed) {
            // insert into cache + maintain size
            ctx.cache.map.insert(
                key,
                CacheEntry {
                    path: out.clone(),
                    last_used_sec: ctx.time_now_sec,
                },
            );
            ctx.cache.lru.push_back((key, ctx.time_now_sec));
            enforce_cache_limits(ctx.time_now_sec, ctx.cfg, ctx.cache);
            return out;
        }

        if allowed_regions.is_none() {
            break;
        }
    }

    Vec::new()
}

fn astar_road_graph(
    ctx: &mut PathfindingCtx<'_>,
    start_idx: usize,
    goal_idx: usize,
    allowed_regions: Option<&[bool]>,
) -> Option<Vec<TilePos>> {
    let w = ctx.graph.width;
    let len = ctx.graph.edges.len();

    // A* over road graph
    let mut came_from: Vec<Option<usize>> = vec![None; len];
    let mut best_g: Vec<u32> = vec![u32::MAX; len];
    let mut heap = BinaryHeap::<HeapState>::new();

    best_g[start_idx] = 0;
    heap.push(HeapState {
        g: 0,
        f: manhattan_idx(start_idx, goal_idx, w),
        idx: start_idx,
    });

    let regions = ctx.regions;
    let is_allowed = |idx: usize| -> bool {
        let Some(mask) = allowed_regions else {
            return true;
        };
        let Some(rg) = regions else {
            return true;
        };
        let Some(rid) = rg.region_id(idx_to_pos(idx, w)) else {
            return true;
        };
        mask.get(rid).copied().unwrap_or(true)
    };

    while let Some(HeapState { g, idx, .. }) = heap.pop() {
        if g != best_g[idx] {
            continue;
        }
        if idx == goal_idx {
            let mut out = Vec::new();
            let mut cur = Some(goal_idx);
            while let Some(ci) = cur {
                out.push(idx_to_pos(ci, w));
                cur = came_from[ci];
            }
            out.reverse();
            return Some(out);
        }

        let mask = ctx.graph.edges[idx];
        if mask == 0 {
            continue;
        }

        // neighbors in W,E,S,N
        let mut push_neighbor = |nidx: usize, step_cost: u32| {
            if !is_allowed(nidx) {
                return;
            }
            let ng = g.saturating_add(step_cost.max(1));
            if ng < best_g[nidx] {
                best_g[nidx] = ng;
                came_from[nidx] = Some(idx);
                let f = ng.saturating_add(manhattan_idx(nidx, goal_idx, w));
                heap.push(HeapState {
                    g: ng,
                    f,
                    idx: nidx,
                });
            }
        };

        if (mask & (1 << 0)) != 0 && idx > 0 {
            push_neighbor(
                idx - 1,
                cost::step_cost_for_edge(cost::StepCostParams {
                    cur_idx: idx,
                    next_idx: idx - 1,
                    move_dir: RoadDir::West,
                    w,
                    cfg: ctx.cfg,
                    traffic: ctx.traffic,
                    grid: ctx.grid,
                    intersections: ctx.intersections,
                }),
            );
        }
        if (mask & (1 << 1)) != 0 && idx + 1 < len {
            push_neighbor(
                idx + 1,
                cost::step_cost_for_edge(cost::StepCostParams {
                    cur_idx: idx,
                    next_idx: idx + 1,
                    move_dir: RoadDir::East,
                    w,
                    cfg: ctx.cfg,
                    traffic: ctx.traffic,
                    grid: ctx.grid,
                    intersections: ctx.intersections,
                }),
            );
        }
        if (mask & (1 << 2)) != 0 && idx >= w {
            push_neighbor(
                idx - w,
                cost::step_cost_for_edge(cost::StepCostParams {
                    cur_idx: idx,
                    next_idx: idx - w,
                    move_dir: RoadDir::South,
                    w,
                    cfg: ctx.cfg,
                    traffic: ctx.traffic,
                    grid: ctx.grid,
                    intersections: ctx.intersections,
                }),
            );
        }
        if (mask & (1 << 3)) != 0 && idx + w < len {
            push_neighbor(
                idx + w,
                cost::step_cost_for_edge(cost::StepCostParams {
                    cur_idx: idx,
                    next_idx: idx + w,
                    move_dir: RoadDir::North,
                    w,
                    cfg: ctx.cfg,
                    traffic: ctx.traffic,
                    grid: ctx.grid,
                    intersections: ctx.intersections,
                }),
            );
        }
    }

    None
}

fn compute_allowed_regions(
    ctx: &PathfindingCtx<'_>,
    start: TilePos,
    goal: TilePos,
) -> Option<Vec<bool>> {
    if !ctx.cfg.enable_hierarchical {
        return None;
    }
    let rg = ctx.regions?;
    if !rg.is_built_for(
        ctx.graph.version,
        ctx.cfg.region_size.max(1),
        ctx.graph.width,
        ctx.graph.height,
    ) {
        return None;
    }

    let start_r = rg.region_id(start)?;
    let goal_r = rg.region_id(goal)?;
    if start_r == goal_r {
        return None;
    }

    let region_path = bfs_region_path(rg, start_r, goal_r)?;
    let pad = ctx.cfg.region_pad.max(0);

    let mut allowed = vec![false; rg.edges.len()];
    for rid in region_path {
        let rx = (rid % rg.regions_w) as i32;
        let ry = (rid / rg.regions_w) as i32;
        for dy in -pad..=pad {
            for dx in -pad..=pad {
                let nx = rx + dx;
                let ny = ry + dy;
                if nx < 0 || ny < 0 {
                    continue;
                }
                let nxu = nx as usize;
                let nyu = ny as usize;
                if nxu >= rg.regions_w || nyu >= rg.regions_h {
                    continue;
                }
                allowed[nyu * rg.regions_w + nxu] = true;
            }
        }
    }

    Some(allowed)
}

fn bfs_region_path(rg: &RegionGraph, start: usize, goal: usize) -> Option<Vec<usize>> {
    let n = rg.edges.len();
    if start >= n || goal >= n {
        return None;
    }
    let mut pred = vec![usize::MAX; n];
    let mut q = VecDeque::new();
    pred[start] = start;
    q.push_back(start);

    while let Some(cur) = q.pop_front() {
        if cur == goal {
            break;
        }
        let mask = rg.edges[cur];
        let x = cur % rg.regions_w;
        let y = cur / rg.regions_w;

        let visit = |nidx: usize, pred: &mut [usize], q: &mut VecDeque<usize>| {
            if pred[nidx] == usize::MAX {
                pred[nidx] = cur;
                q.push_back(nidx);
            }
        };

        if (mask & (1 << 0)) != 0 && x > 0 {
            visit(cur - 1, &mut pred, &mut q);
        }
        if (mask & (1 << 1)) != 0 && x + 1 < rg.regions_w {
            visit(cur + 1, &mut pred, &mut q);
        }
        if (mask & (1 << 2)) != 0 && y > 0 {
            visit(cur - rg.regions_w, &mut pred, &mut q);
        }
        if (mask & (1 << 3)) != 0 && y + 1 < rg.regions_h {
            visit(cur + rg.regions_w, &mut pred, &mut q);
        }
    }

    if pred[goal] == usize::MAX {
        return None;
    }
    let mut path = Vec::new();
    let mut cur = goal;
    path.push(cur);
    while cur != start {
        cur = pred[cur];
        path.push(cur);
    }
    path.reverse();
    Some(path)
}

fn enforce_cache_limits(time_now_sec: f64, cfg: &PathfindingConfig, cache: &mut PathCache) {
    // TTL purge (approximate, front-biased)
    while let Some((key, used)) = cache.lru.front().copied() {
        if time_now_sec - used <= cfg.cache_ttl_secs {
            break;
        }
        cache.lru.pop_front();
        if let Some(e) = cache.map.get(&key)
            && (e.last_used_sec - used).abs() < f64::EPSILON
        {
            cache.map.remove(&key);
        }
    }

    // Capacity purge (approximate LRU)
    while cache.map.len() > cfg.cache_capacity {
        let Some((key, used)) = cache.lru.pop_front() else {
            break;
        };
        if let Some(e) = cache.map.get(&key)
            && (e.last_used_sec - used).abs() < f64::EPSILON
        {
            cache.map.remove(&key);
        }
    }
}

/// Async pathfinding request ID
#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
pub struct PathRequestId(pub u64);

/// Async pathfinding request
#[derive(Debug, Clone)]
pub struct PathRequest {
    pub id: PathRequestId,
    pub start: TilePos,
    pub goal: TilePos,
    pub priority: i32, // Higher = more important
}

/// Async pathfinding result
#[derive(Debug)]
pub struct PathResult {
    pub request_id: PathRequestId,
    pub path: Option<Vec<TilePos>>,
}

/// Queue of pending pathfinding requests
#[derive(Resource, Default)]
pub struct PathRequestQueue {
    next_id: u64,
    requests: Vec<PathRequest>,
    processing: HashMap<PathRequestId, PathRequest>,
}

/// Results of completed pathfinding requests
#[derive(Resource, Default)]
pub struct PathResultQueue {
    results: Vec<PathResult>,
}

/// Plugin for async pathfinding system
pub struct AsyncPathfindingPlugin;

impl Plugin for AsyncPathfindingPlugin {
    fn build(&self, app: &mut App) {
        app.init_resource::<PathRequestQueue>()
            .init_resource::<PathResultQueue>()
            .add_systems(FixedUpdate, process_pathfinding_requests);
    }
}

impl PathRequestQueue {
    /// Submit a new pathfinding request
    pub fn submit(&mut self, start: TilePos, goal: TilePos, priority: i32) -> PathRequestId {
        let id = PathRequestId(self.next_id);
        self.next_id = self.next_id.wrapping_add(1);

        let request = PathRequest {
            id,
            start,
            goal,
            priority,
        };
        self.requests.push(request);
        id
    }

    /// Get next request to process (highest priority first)
    pub fn pop_next(&mut self) -> Option<PathRequest> {
        if self.requests.is_empty() {
            return None;
        }

        // Find highest priority request
        let mut best_idx = 0;
        let mut best_priority = self.requests[0].priority;

        for (i, req) in self.requests.iter().enumerate().skip(1) {
            if req.priority > best_priority {
                best_priority = req.priority;
                best_idx = i;
            }
        }

        let request = self.requests.swap_remove(best_idx);
        self.processing.insert(request.id, request.clone());
        Some(request)
    }

    /// Mark request as completed
    pub fn complete(&mut self, id: PathRequestId) {
        self.processing.remove(&id);
    }

    /// Get pending request count
    pub fn pending_count(&self) -> usize {
        self.requests.len()
    }

    /// Get processing request count
    pub fn processing_count(&self) -> usize {
        self.processing.len()
    }
}

impl PathResultQueue {
    /// Add completed pathfinding result
    pub fn push(&mut self, result: PathResult) {
        self.results.push(result);
    }

    /// Get all pending results
    pub fn drain(&mut self) -> std::vec::Drain<PathResult> {
        self.results.drain(..)
    }
}

/// Process pathfinding requests with budget per tick
fn process_pathfinding_requests(
    time: Res<Time>,
    cfg: Res<PathfindingConfig>,
    mut cache: ResMut<PathCache>,
    graph: Res<RoadGraph>,
    regions: Option<Res<RegionGraph>>,
    traffic: Option<Res<TrafficOccupancy>>,
    grid: Res<MapGrid>,
    intersections: Res<IntersectionIndex>,
    mut request_queue: ResMut<PathRequestQueue>,
    mut result_queue: ResMut<PathResultQueue>,
) {
    // Budget: process up to 8 requests per tick to avoid stalls
    const MAX_REQUESTS_PER_TICK: usize = 8;
    let mut processed = 0;

    // Default traffic occupancy for when none is available
    let default_traffic = TrafficOccupancy::default();

    while processed < MAX_REQUESTS_PER_TICK {
        let Some(request) = request_queue.pop_next() else {
            break;
        };

        // Create pathfinding context for this request
        let mut ctx = PathfindingCtx {
            time_now_sec: time.elapsed_secs_f64(),
            cfg: &cfg,
            cache: &mut cache,
            graph: &graph,
            regions: regions.as_deref(),
            traffic: traffic.as_deref().unwrap_or(&default_traffic),
            grid: &grid,
            intersections: &intersections,
        };

        // Process the pathfinding request
        let path = find_road_path_cached(&mut ctx, request.start, request.goal);

        // Store result
        result_queue.push(PathResult {
            request_id: request.id,
            path: Some(path),
        });

        // Mark as completed
        request_queue.complete(request.id);
        processed += 1;
    }
}

File: packages/sim/src/transport/pathfinding.ts
// Port of crates/simcity_sim/src/game/transport/pathfinding/{mod,astar,cost,cache,regions}.rs:
// road-A* over the road graph with congestion-aware integer costs, a region pre-pass and a
// TTL + approximate-LRU path cache. Edge costs are computed in f32 exactly as Rust does.
import type { RoadDir, TilePos } from '../commands';
import type { IntersectionIndex } from '../intersections/index';
import type { MapGrid } from '../map/grid';
import { capacityPerLaneTile, dirLeft, dirRight, roadDesirability, roadSpeedLimit } from '../map/roads';
import type { TrafficOccupancy } from '../traffic/occupancy';
import { OpenSet } from './openSet';
import type { RegionGraph } from './regionGraph';
import type { RoadGraph } from './roadGraph';

const f32 = Math.fround;
const U32_MAX = 0xffff_ffff;
const MAX_CACHED_PATH = 1000;
/** Seconds a vehicle waits at a light on average, priced into the edge. */
const AVG_LIGHT_WAIT_SECS = 5;

export interface PathfindingConfig {
  cacheCapacity: number;
  cacheTtlSecs: number;
  /** `k` in `w = base_cost * (1 + k * congestion)`, `f32`. */
  congestionK: number;
  /** Clamp for occupancy / capacity, `f32`. */
  congestionMax: number;
  laneChangePenalty: number;
  turnPenalty: number;
  /** Float weights -> integer A* costs, `f32`. */
  costScale: number;
  enableHierarchical: boolean;
  regionSize: number;
  regionPad: number;
}

/** `PathfindingConfig::default()` (and assets/config/pathfinding.ron). */
export function defaultPathfindingConfig(): PathfindingConfig {
  return {
    cacheCapacity: 4096,
    cacheTtlSecs: 10,
    congestionK: 2,
    congestionMax: 2,
    laneChangePenalty: 40,
    turnPenalty: 80,
    costScale: 1000,
    enableHierarchical: true,
    regionSize: 16,
    regionPad: 1,
  };
}

interface CacheEntry {
  readonly path: readonly TilePos[];
  lastUsedSec: number;
}

export class PathCache {
  version = 0;
  map = new Map<string, CacheEntry>();
  /** `(key, last_used_sec at push)`; duplicates allowed, stale entries skipped on purge. */
  lru: Array<readonly [string, number]> = [];

  clear(): void {
    this.map.clear();
    this.lru = [];
  }
}

export interface PathfindingCtx {
  timeNowSec: number;
  cfg: PathfindingConfig;
  cache: PathCache;
  graph: RoadGraph;
  regions: RegionGraph | null;
  traffic: TrafficOccupancy;
  grid: MapGrid;
  intersections: IntersectionIndex;
}

const pathKey = (start: TilePos, goal: TilePos, version: number) =>
  `${start.x},${start.y}|${goal.x},${goal.y}|${version}`;

function enforceCacheLimits(timeNowSec: number, cfg: PathfindingConfig, cache: PathCache): void {
  let head = 0;
  // TTL purge (approximate, front-biased).
  while (head < cache.lru.length) {
    const [key, used] = cache.lru[head]!;
    if (timeNowSec - used <= cfg.cacheTtlSecs) break;
    head++;
    const entry = cache.map.get(key);
    if (entry !== undefined && Math.abs(entry.lastUsedSec - used) < Number.EPSILON) cache.map.delete(key);
  }
  // Capacity purge (approximate LRU).
  while (cache.map.size > cfg.cacheCapacity && head < cache.lru.length) {
    const [key, used] = cache.lru[head]!;
    head++;
    const entry = cache.map.get(key);
    if (entry !== undefined && Math.abs(entry.lastUsedSec - used) < Number.EPSILON) cache.map.delete(key);
  }
  if (head > 0) cache.lru = cache.lru.slice(head);
}

/** `step_cost_for_edge`: integer cost of moving from `curIdx` to `nextIdx`. */
function stepCostForEdge(ctx: PathfindingCtx, curIdx: number, nextIdx: number, moveDir: RoadDir): number {
  const { grid, cfg } = ctx;
  const w = ctx.graph.width;
  const curPos = { x: curIdx % w, y: Math.floor(curIdx / w) };
  const nextPos = { x: nextIdx % w, y: Math.floor(nextIdx / w) };
  const cur = grid.get(curPos)?.road;
  const next = grid.get(nextPos)?.road;
  const kind = next?.kind ?? 'None';

  const speed = f32(Math.max(roadSpeedLimit(kind), 1));
  const capacity = f32(Math.max(capacityPerLaneTile(kind), 1));
  const desirability = f32(Math.max(roadDesirability(kind), f32(0.1)));
  const occupancy = f32(ctx.traffic.perTickVehicles[nextIdx] ?? 0);
  const congestion = Math.min(Math.max(f32(occupancy / capacity), 0), f32(Math.max(cfg.congestionMax, 0)));

  const travelTime = f32(1 / speed);
  const desirabilityFactor = f32(1 / desirability);
  const baseCost = f32(travelTime * desirabilityFactor);
  const congestionFactor = f32(1 + f32(cfg.congestionK * congestion));
  const costScale = f32(Math.max(cfg.costScale, 1));
  const raw = f32(f32(baseCost * congestionFactor) * costScale);

  let penalty = 0;
  if (cur !== undefined && next !== undefined && cur.dir !== 'None' && next.dir !== 'None') {
    const perpendicular = moveDir === dirLeft(cur.dir) || moveDir === dirRight(cur.dir);
    if (perpendicular && next.dir === cur.dir) penalty = f32(penalty + Math.max(cfg.laneChangePenalty, 0));
    else if (perpendicular && next.dir === moveDir) penalty = f32(penalty + Math.max(cfg.turnPenalty, 0));
  }

  const lightPenalty = ctx.intersections.hasTrafficLightAt(nextPos) ? f32(AVG_LIGHT_WAIT_SECS * costScale) : 0;
  const total = Math.max(f32(f32(raw + penalty) + lightPenalty), 1);
  return Math.min(Math.trunc(total), U32_MAX);
}

function manhattanIdx(a: number, b: number, w: number): number {
  return Math.abs((a % w) - (b % w)) + Math.abs(Math.floor(a / w) - Math.floor(b / w));
}

function bfsRegionPath(rg: RegionGraph, start: number, goal: number): number[] | undefined {
  const n = rg.edges.length;
  if (start >= n || goal >= n) return undefined;
  const pred = new Int32Array(n).fill(-1);
  pred[start] = start;
  const queue = [start];
  for (let head = 0; head < queue.length; head++) {
    const cur = queue[head]!;
    if (cur === goal) break;
    const mask = rg.edges[cur]!;
    const x = cur % rg.regionsW;
    const y = Math.floor(cur / rg.regionsW);
    const visit = (nidx: number) => {
      if (pred[nidx] === -1) {
        pred[nidx] = cur;
        queue.push(nidx);
      }
    };
    if ((mask & 1) !== 0 && x > 0) visit(cur - 1);
    if ((mask & 2) !== 0 && x + 1 < rg.regionsW) visit(cur + 1);
    if ((mask & 4) !== 0 && y > 0) visit(cur - rg.regionsW);
    if ((mask & 8) !== 0 && y + 1 < rg.regionsH) visit(cur + rg.regionsW);
  }
  if (pred[goal] === -1) return undefined;
  const path = [goal];
  let cur = goal;
  while (cur !== start) {
    cur = pred[cur]!;
    path.push(cur);
  }
  return path.reverse();
}

/** Regions the low-level search may enter: the BFS region path padded by `regionPad`. */
function computeAllowedRegions(ctx: PathfindingCtx, start: TilePos, goal: TilePos): Uint8Array | undefined {
  const rg = ctx.regions;
  if (!ctx.cfg.enableHierarchical || rg === null) return undefined;
  if (!rg.isBuiltFor(ctx.graph.version, Math.max(ctx.cfg.regionSize, 1), ctx.graph.width, ctx.graph.height)) {
    return undefined;
  }
  const startR = rg.regionId(start);
  const goalR = rg.regionId(goal);
  if (startR === undefined || goalR === undefined || startR === goalR) return undefined;
  const regionPath = bfsRegionPath(rg, startR, goalR);
  if (regionPath === undefined) return undefined;

  const pad = Math.max(ctx.cfg.regionPad, 0);
  const allowed = new Uint8Array(rg.edges.length);
  for (const rid of regionPath) {
    const rx = rid % rg.regionsW;
    const ry = Math.floor(rid / rg.regionsW);
    for (let dy = -pad; dy <= pad; dy++) {
      for (let dx = -pad; dx <= pad; dx++) {
        const nx = rx + dx;
        const ny = ry + dy;
        if (nx < 0 || ny < 0 || nx >= rg.regionsW || ny >= rg.regionsH) continue;
        allowed[ny * rg.regionsW + nx] = 1;
      }
    }
  }
  return allowed;
}

function astarRoadGraph(
  ctx: PathfindingCtx,
  startIdx: number,
  goalIdx: number,
  allowedRegions: Uint8Array | undefined,
): TilePos[] | undefined {
  const { graph } = ctx;
  const w = graph.width;
  const len = graph.edges.length;
  const cameFrom = new Int32Array(len).fill(-1);
  const bestG = new Uint32Array(len).fill(U32_MAX);
  const open = new OpenSet();

  bestG[startIdx] = 0;
  open.push(manhattanIdx(startIdx, goalIdx, w), 0, startIdx);

  const isAllowed = (idx: number): boolean => {
    if (allowedRegions === undefined || ctx.regions === null) return true;
    const rid = ctx.regions.regionId({ x: idx % w, y: Math.floor(idx / w) });
    return rid === undefined || rid >= allowedRegions.length || allowedRegions[rid] === 1;
  };

  while (open.size > 0) {
    const [g, idx] = open.pop();
    if (g !== bestG[idx]) continue;
    if (idx === goalIdx) {
      const out: TilePos[] = [];
      for (let cur = goalIdx; cur !== -1; cur = cameFrom[cur]!) out.push({ x: cur % w, y: Math.floor(cur / w) });
      return out.reverse();
    }
    const mask = graph.edges[idx]!;
    if (mask === 0) continue;

    const pushNeighbour = (nidx: number, moveDir: RoadDir) => {
      if (!isAllowed(nidx)) return;
      const step = Math.max(stepCostForEdge(ctx, idx, nidx, moveDir), 1);
      const ng = Math.min(g + step, U32_MAX);
      if (ng < bestG[nidx]!) {
        bestG[nidx] = ng;
        cameFrom[nidx] = idx;
        open.push(Math.min(ng + manhattanIdx(nidx, goalIdx, w), U32_MAX), ng, nidx);
      }
    };
    if ((mask & 1) !== 0 && idx > 0) pushNeighbour(idx - 1, 'West');
    if ((mask & 2) !== 0 && idx + 1 < len) pushNeighbour(idx + 1, 'East');
    if ((mask & 4) !== 0 && idx >= w) pushNeighbour(idx - w, 'South');
    if ((mask & 8) !== 0 && idx + w < len) pushNeighbour(idx + w, 'North');
  }
  return undefined;
}

function findRoadPath(ctx: PathfindingCtx, start: TilePos, goal: TilePos): TilePos[] {
  if (start.x === goal.x && start.y === goal.y) return [start];
  const { graph } = ctx;
  if (graph.edges.length === 0 || graph.width === 0) return [];
  const w = graph.width;
  const len = graph.edges.length;
  // Row-major indices from the graph width, bounds-checked only against the whole array (as Rust).
  const startIdx = start.y * w + start.x;
  const goalIdx = goal.y * w + goal.x;
  if (startIdx < 0 || goalIdx < 0 || startIdx >= len || goalIdx >= len) return [];
  if (graph.edges[startIdx] === 0) return [];
  return astarRoadGraph(ctx, startIdx, goalIdx, computeAllowedRegions(ctx, start, goal)) ?? [];
}

/**
 * A road path from the cache or a fresh A*. Empty when the graph is not built, start is not a road
 * node, or no path exists.
 */
export function findRoadPathCached(ctx: PathfindingCtx, start: TilePos, goal: TilePos): TilePos[] {
  const { cache, graph } = ctx;
  if (cache.version !== graph.version) {
    cache.clear();
    cache.version = graph.version;
  }
  const key = pathKey(start, goal, graph.version);
  const hit = cache.map.get(key);
  if (hit !== undefined) {
    hit.lastUsedSec = ctx.timeNowSec;
    cache.lru.push([key, ctx.timeNowSec]);
    return hit.path.slice();
  }

  const path = findRoadPath(ctx, start, goal);
  if (path.length >= 2 && path.length <= MAX_CACHED_PATH) {
    cache.map.set(key, { path: path.slice(), lastUsedSec: ctx.timeNowSec });
    cache.lru.push([key, ctx.timeNowSec]);
    enforceCacheLimits(ctx.timeNowSec, ctx.cfg, cache);
  }
  return path;
}
